Rotate Encryption Keys for RKE2

jakefhyde opened the pull request:
## Description

rke2/k3s support rotating encryption keys for secrets via the cli, which is required for parity with rke1 so v2provisioning can go GA. This is implemented by first calling the correct command on an elected control plane node (in this case, we will use the init node), and then restarting the init node. After the init node has been restarted, all other nodes can be safely restarted. This is according to our [official docs on the rotating encryption secrets with k3s](https://rancher.com/docs/k3s/latest/en/security/secrets_encryption/#high-availability-encryption-key-rotation). If the nodes are out of sync (the usual ran commands/restarted nodes), the planner will alternate between restarting the init node and restarting the other control plane nodes until all nodes match. 

Related Issue: https://github.com/rancher/rancher/issues/35436

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+		return nil
+	}
+
+	currentInitNodes := collect(clusterPlan, isInitNode)
```
You shouldn't need to do this, there is a helper function to find the init node in `initnode.go` of the planner

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+		}
+	}
+
+	// init node changed in between get and list
```
If the init node changes between a get and list we have a huge problem

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+	}
+	initEntry := currentInitNodes[0]
+
+	clientset, err := rke2.ClientSetFromCache(initEntry.Machine.Namespace, initEntry.Machine.Spec.ClusterName, r.rancherClusterCache, r.kubeconfigManager)
```
I'm not convinced we need to/should be generating a clientset in the planner that hits the downstream cluster... can we not scrape data from plan outputs to determine the status of the rotation?

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+		}
+	}
+
+	if phase == "reencrypt_request" || phase == "reencrypt_active" {
```
these should probably be defined as constants

Oats87 reviewed (changes requested):
Stylistically, this is a departure from the established "format" of the way functionality has been implemented in the v2prov planner.

I would be more comfortable if we had uniformity within the planner, the etcdrestore portion is a great example of the way functionality has been implemented today, and makes a very confusing process somewhat less confusing (with adequate comments).

I need to re-familiarize myself with the secrets-encryption feature in Rancher before I can properly scrutinize the procedural logic here; however, I will say that it is extremely important that the secrets encryption process is leaktight/fool proof, otherwise we can end up in very bad situations with non-accessible secrets (thus requiring an etcd snapshot restore etc.)

On first glance, the procedural logic seems to make sense.

Oats87 commented:
I'm not sure that these functions warrant being hidden under their own struct. I had to end up moving the etcd create/etcd restore structs out to being first class in the planner due to the non-accessible variables/functions/etc. and I'm seeing that this might be necessary for this feature as well

paynejacob commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+		return err
+	}
+
+	nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
```
why not use collect for this?

thedadams replied:
Collect is for our plan stuff, not the `v1.Node`s.

paynejacob replied:
ah missed that

thedadams commented on pkg/apis/rke.cattle.io/v1/controlplane.go:
```diff
@@ -44,15 +45,27 @@ var (
+	RotateEncryptionKeysPhaseStart         RotateEncryptionKeysPhase = "Start"
+	RotateEncryptionKeysPhaseApply         RotateEncryptionKeysPhase = "Apply"
+	RotateEncryptionKeysPhaseRestart       RotateEncryptionKeysPhase = "Restart"
+	RotateEncryptionKeysPhaseRestartOthers RotateEncryptionKeysPhase = "RestartOthers"
```
This one could be `RotateEncryptionKeysPhaseRestartFollowers` and the one above could be `RotateEncrpytionKeysPhaseRestartLeader`

thedadams commented on pkg/controllers/provisioningv2/rke2/provisioningcluster/template.go:
```diff
@@ -491,6 +491,7 @@ func rkeControlPlane(cluster *rancherv1.Cluster) (*rkev1.RKEControlPlane, error)
 			ETCDSnapshotRestore:      cluster.Spec.RKEConfig.ETCDSnapshotRestore.DeepCopy(),
 			ETCDSnapshotCreate:       cluster.Spec.RKEConfig.ETCDSnapshotCreate.DeepCopy(),
 			RotateCertificates:       cluster.Spec.RKEConfig.RotateCertificates.DeepCopy(),
+			RotateEncryptionKeys:     cluster.Spec.RKEConfig.RotateEncryptionKeys.DeepCopy(),
```
I know this isn't really part of your PR, but can we put an `rkeConfig := cluster.Spec.RKEconfig.DeepCopy()` above and use the pieces here without all the `DeepCopy` calls?

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+	}
+}
+
+func (r *encryptionKeyRotation) updatePlan(cp *rkev1.RKEControlPlane, entry *planEntry, node *v1.Node, instructions ...plan.Instruction) error {
```
There is already an `UpdatePlan` method in `store.go`. Can that be used here?

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,272 @@
+			}
+			// restart nodes that are not in sync
+			for _, entry := range collect(clusterPlan, isControlPlaneAndNotInitNode) {
+				if k == phase {
+					continue
+				}
```
This is checked above. Can this be removed?

thedadams commented on pkg/controllers/provisioningv2/rke2/common.go:
```diff
@@ -206,3 +213,17 @@ func CopyMapWithExcludes(destination map[string]string, source map[string]string
 	}
 }
+
+func ClientSetFromCache(namespace, clusterName string, rancherClusterCache ranchercontrollers.ClusterCache, kubeconfigManager *kubeconfig.Manager) (*kubernetes.Clientset, error) {
```
I believe this code is also used in `machinenodelookup`. Can we use this function there too?

Oats87 commented on pkg/controllers/provisioningv2/rke2/common.go:
```diff
@@ -11,6 +11,7 @@ import (
 	provv1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
 	"github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1/plan"
 	capicontrollers "github.com/rancher/rancher/pkg/generated/controllers/cluster.x-k8s.io/v1beta1"
+
```
nit: remove newline

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+)
+
+const (
+	Start             = "start"
```
Given that these constants are under the planner package, can we rename them to indicate that they're for encryption key rotation to help prevent confusion?

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+	}
+
+	if shouldRestartEncryptionKeyRotation(cp) {
+		clusterPlan.Cluster.Annotations[capi.PausedAnnotation] = "true"
```
Are we safe to start manipulating the passed in `clusterPlan` object without a `DeepCopy`?

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+	if err != nil {
+		return err
+	}
+	if leader == nil {
```
An error message/log message for this would probably be helpful, otherwise we may run into situations where we don't know why the planner isn't progressing.

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+	return generic.ErrSkip
+}
+
+func (p *Planner) enqueueAndSkip(cp *rkev1.RKEControlPlane) error {
```
Can we put this in a more "central" location for the planner?

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+	}
+}
+
+func (p *Planner) applyAndWait(cp *rkev1.RKEControlPlane, entry *planEntry, node plan.NodePlan) error {
```
I'm not sure why this function needs to be defined this way, you can probably just directly call `assignAndCheckPlan`

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+			return p.enqueueIfErrWaiting(cp, err)
+		}
+
+		logrus.Infof("Rotate encryption keys cluster [%s]: running apply command: [%s]", cp.Spec.ClusterName, apply.Args[1]) // todo Debug
```
is this supposed to be a `logrus.Debugf`?

Oats87 commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+	if err != nil {
+		w := ErrWaiting("")
+		if errors.As(err, &w) {
+			logrus.Infof("Rotate encryption keys cluster [%s]: %s", cp.Spec.ClusterName, w.Error()) // todo Debug
```
nit: debugf?

Oats87 commented on pkg/provisioningv2/rke2/planner/initnode.go:
```diff
@@ -65,35 +65,35 @@ func (p *Planner) findAndDesignateFixedInitNode(rkeControlPlane *rkev1.RKEContro
 }
 
 // findInitNode searches the given cluster for the init node. It returns a bool which is whether an init node was
 // found, the init node join URL, and an error for a few conditions, i.e. if multiple init nodes were found or if there
```
Can you update the godoc comment to reflect that the function is now returning the node entry as well for the found init node?

Oats87 commented on pkg/provisioningv2/rke2/planner/planner.go:
```diff
@@ -216,7 +225,12 @@ func (p *Planner) Process(controlPlane *rkev1.RKEControlPlane) error {
+		return err
+	}
+
+	err = p.RotateEncryptionKeys(controlPlane, plan)
```
```suggestion
	if err = p.RotateEncryptionKeys(controlPlane, plan); err != nil {
	    return err
	}
```

Oats87 reviewed (changes requested):
On first pass, overall this looks great; however, I think in its current state merging it may cause confusion for the next person to stumble across the procedural logic here.

jakefhyde commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,450 @@
+	case windows:
+		lastKnownLeaderPhaseEnv = fmt.Sprintf("$env:LAST_KNOWN_LEADER_PHASE=%s", leaderPhase)
+	default:
+		lastKnownLeaderPhaseEnv = fmt.Sprintf("LAST_KNOWN_LEADER_PHASE=%s", leaderPhase)
```
Updated comment for clarity.

jakefhyde commented:
The machine health checks definitely interfered with some things (also the default unhealthy node timeout being 0s caused some issues, going to open a UI issue for that), addressed review comments but need to bump down enqueue interval from 10 seconds to 2, and try `killall.sh` again.

Oats87 commented on pkg/provisioningv2/rke2/planner/planner.go:
```diff
@@ -110,18 +116,19 @@ type roleFilter func(*planEntry) bool
 	ctx                           context.Context
 	store                         *PlanStore
-	rkeControlPlanes              rkecontrollers.RKEControlPlaneClient
+	rkeControlPlanes              rkecontrollers.RKEControlPlaneController
```
Out of curiosity, why are we changing to the `RKEControlPlaneController` here instead of the `RKEControlPlaneClient`?

jakefhyde replied:
We can only enqueue with the controller, (which also is a client).

Oats87 commented on pkg/provisioningv2/rke2/planner/instructions.go:
```diff
@@ -54,7 +54,7 @@ func generateInstallInstruction(controlPlane *rkev1.RKEControlPlane, entry *plan
 // addInstallInstructionWithRestartStamp will generate an instruction and append it to the node plan that executes the `run.sh` or `run.ps1`
 // from the installer image based on the control plane configuration. It will generate a restart stamp based on the
-// passed in configuration to determine whether it needs to start/restart the service being managed.
+// passed in configuration to determine whether it needs to encryptionKeyRotationStart/restart the service being managed.
```
nit: does this change to the comment make sense?

Oats87 commented on pkg/controllers/provisioningv2/rke2/provisioningcluster/controller.go:
```diff
@@ -279,7 +279,7 @@ func (h *handler) getRKEControlPlaneForCluster(cluster *rancherv1.Cluster) (*rke
 
 	cp, err := h.rkeControlPlane.Get(capiCluster.Spec.ControlPlaneRef.Namespace, capiCluster.Spec.ControlPlaneRef.Name)
-	if apierror.IsNotFound(err) && cluster.DeletionTimestamp == nil {
+	if apierror.IsNotFound(err) && cluster.DeletionTimestamp != nil {
```
nice catch

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,440 @@
+		return p.enqueueIfErrWaiting(cp, err)
+	}
+	if phase == EncryptionKeyRotationReencryptRequest || phase == EncryptionKeyRotationReencryptActive {
+		if periodicPhase != EncryptionKeyRotationReencryptRequest && periodicPhase != EncryptionKeyRotationReencryptActive && periodicPhase != EncryptionKeyRotationReencryptFinished {
```
```suggestion
		if periodicPhase != phase && periodicPhase != EncryptionKeyRotationReencryptFinished 
```

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,440 @@
+}
+
+func (p *Planner) encryptionKeyRotationUpdateControlPlanePhase(cp *rkev1.RKEControlPlane, phase rkev1.RotateEncryptionKeysPhase) error {
+	cp.Status.RotateEncryptionKeysPhase = phase
```
Add a check here to see if we need to update the `phase`.

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,440 @@
+	}, nil
+}
+
+func encryptionKeyRotationStatusInstruction(cp *rkev1.RKEControlPlane) plan.OneTimeInstruction {
```
I think `encryptionKeyRotationStatusOneTimeInstruction` is a more descriptive name.

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,440 @@
+	}
+}
+
+func encryptionKeyRotationFindApplyCommand(entry *planEntry) (plan.OneTimeInstruction, error) {
```
Can we add a comment to this to explain what it does? I also think a better name could be found here, but I am having a hard time coming up with a short one.

thedadams commented on pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go:
```diff
@@ -0,0 +1,440 @@
+}
+
+// encryptionKeyRotationExpectedPhase returns the expected secrets-encrypt phase extracted from the leader's apply command present in the plan.
+func encryptionKeyRotationExpectedPhase(leaderPhase string, leader *planEntry) (bool, error) {
```
The name of this function implies that it returns the expected status. It seems that it returns whether the phase is what is expected based on the command.
```suggestion
func encryptionKeyRotationPhaseIsExpectedFromCommand(leaderPhase string, leader *planEntry) (bool, error) {
```

jakefhyde commented on pkg/provisioningv2/rke2/planner/initnode.go:
```diff
@@ -39,19 +39,19 @@ func (p *Planner) setInitNodeMark(entry *planEntry) error {
 // findAndDesignateFixedInitNode is used for rancherd where an exact machine (determined by labeling the
 // rkecontrolplane object) is desired to be the init node
-func (p *Planner) findAndDesignateFixedInitNode(rkeControlPlane *rkev1.RKEControlPlane, plan *plan.Plan) (bool, string, error) {
+func (p *Planner) findAndDesignateFixedInitNode(rkeControlPlane *rkev1.RKEControlPlane, plan *plan.Plan) (bool, string, *planEntry, error) {
```
Unfortunately I wasn't able to implement this change, because `findInitNode` can return true with a `nil` init node to signify that there is an existing init node, with an empty join URL.
